controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -0,0 +1,349 @@
+        // try creating txn commit list first. if node already exists and doesnt match the processing in the event, throw operation not allowed.
+        // This will result in event being posted back in the stream and retried later. Generally if a transaction commit starts, it will come to
+        // an end.. but during failover, once we have created the node, we are guaranteed that it will be only that transaction that will be getting
+        // committed at that time.
+        return streamMetadataStore.getState(scope, stream, true, context, executor)
+                .thenCompose(state -> {
+                    CompletableFuture<List<UUID>> txnListFuture = createRecordAndGetCommitTxnList(scope, stream, txnEpoch, context);
+
+                    CompletableFuture<Void> commitFuture = txnListFuture
+                            .thenCompose(txnList -> {
+                                if (txnList == null) {
+                                    // reset state conditionally in case we were left with stale committing state from a previous execution
+                                    // that died just before updating the state back to ACTIVE but after having completed all the work.
+                                    return streamMetadataStore.resetStateConditionally(scope, stream, State.COMMITTING_TXN, context, executor);
+                                } else {
+                                    // Once state is set to committing, we are guaranteed that this will be the only processing that can happen on the stream
+                                    // and we can proceed with committing outstanding transactions collected in the txnList step.
+                                    CompletableFuture<Void> future;
+                                    // if state is sealing, we should continue with commit so that we allow for completion of transactions
+                                    // in commit state.
+                                    if (state.equals(State.SEALING)) {
+                                        future = new CompletableFuture<>();
+                                    } else {
+                                        // In normal course set the state to committing before proceeding.
+                                        // If we are unable to set the state to COMMITTING_TXN, it will get OPERATION_NOT_ALLOWED
+                                        // and the processing will be retried later.
+                                        future = Futures.toVoid(streamMetadataStore.setState(scope, stream, State.COMMITTING_TXN, context, executor));
+                                    }
+
+                                    // Note: since we have set the state to COMMITTING_TXN (or it was already sealing), the active epoch that we fetch now
+                                    // cannot change until we perform rolling txn. TxnCommittingRecord ensures no other rollingTxn
```

Comment:
I don't know what `TxnCommittingRecord` is. I searched for it and there is a single occurrence.

Reply:
This is the list of transactions that we collect for committing as part of this processing. 
Its a list which is persisted in metadata store and upon failure and retry, the entire list is replayed in the same order in idempotent fashion.